File: hibernate-scala-examples/src/main/scala/org/hibernate/examples/model/TreeNodePosition.scala
package org.hibernate.examples.model

import javax.persistence.{Embeddable, Column}
import org.hibernate.examples.utils.{ToStringHelper, Hashs}

/**
 * kr.debop4s.data.model.TreeNodePosition
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2013. 12. 24. 오후 4:03
 */
@SerialVersionUID(3455568346636164669L)
@Embeddable
case class TreeNodePosition(var level: Int = 0, var order: Int = 0) extends ValueObject {

    @Column(name = "treeLevel")
    def getLevel: Int = level

    @Column(name = "treeOrder")
    def getOrder: Int = order

    def setPosition(level: Int, order: Int) {
        this.level = level
        this.order = order
    }

    override def hashCode(): Int = Hashs.compute(level, order)

    override protected def buildStringHelper: ToStringHelper =
        super.buildStringHelper
            .add("level", level)
            .add("order", order)
}

File: hibernate-examples/src/main/java/org/hibernate/examples/model/TreeNodePosition.java
package org.hibernate.examples.model;

import lombok.Getter;
import lombok.Setter;
import org.hibernate.examples.utils.HashTool;
import org.hibernate.examples.utils.ToStringHelper;

import javax.persistence.Column;
import javax.persistence.Embeddable;

/**
 * 계층형 구조 엔티티의 TREE 상에서의 위치를 나타냅니다.
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2013. 11. 27. 오후 4:30
 */
@Embeddable
@Getter
@Setter
public class TreeNodePosition extends AbstractValueObject {

  @Column(name = "treeLevel")
  private Integer level;

  @Column(name = "treeOrder")
  private Integer order;

  public TreeNodePosition() {
    this(0, 0);
  }

  public TreeNodePosition(int level, int order) {
    this.level = level;
    this.order = order;
  }

  public TreeNodePosition(TreeNodePosition src) {
    this(src.level, src.order);
  }

  public void setPosition(int level, int order) {
    this.level = level;
    this.order = order;
  }

  @Override
  public int hashCode() {
    return HashTool.compute(level, order);
  }

  @Override
  public ToStringHelper buildStringHelper() {
    return super.buildStringHelper()
        .add("level", level)
        .add("order", order);
  }

  private static final long serialVersionUID = -5552929284039433528L;
}
